### PR TITLE
fast subinplace

### PR DESCRIPTION
**Pull request by bqpd**
after profiling @mjburton11's code.

depends on #854

### CODE REVIEW

**Comment by bqpd** [comment]
@whoburg, ready for review (just look at the last commit)

**Comment by bqpd** [comment]
@whoburg, ready for review

File: gpkit/keydict.py [hunk 1]
```diff
@@ -62,7 +62,7 @@ def parse_and_index(self, key):
                                          " .variables_byname('%s') will show"
                                          " which variables it may refer to."
                                          % (key, key))
-                else:
+                elif key != "filter":
                     raise KeyError("key '%s' does not refer to any varkey in"
                                    " this ConstraintSet" % key)
```
Comment (whoburg) [inline gpkit/keydict.py]: is `filter` used internally or user-facing?

File: gpkit/constraints/set.py [hunk 2]
```diff
@@ -183,7 +183,7 @@ def subinplace(self, subs):
                     if k not in constraint.varkeys.keymap:
                         continue
                     keys = constraint.varkeys.keymap[k]
-                    if k.shape and isinstance(v, Iterable):
+                    if isinstance(v, Iterable) and k.shape:
```
Comment (whoburg) [inline gpkit/constraints/set.py]: speedup? interesting, I'm surprised this is faster.

File: gpkit/keydict.py [hunk 3]
```diff
@@ -44,6 +44,8 @@ def __init__(self, *args, **kwargs):
     def update(self, *args, **kwargs):
         "Iterates through the dictionary created by args and kwargs"
         for k, v in dict(*args, **kwargs).items():
+            if hasattr(v, "copy"):
```
Comment (whoburg) [inline gpkit/keydict.py]: inline comment might help here -- we don't need a copy unless v provides the method?

File: gpkit/tests/helpers.py [hunk 4]
```diff
@@ -47,7 +47,8 @@ class that inherits from `unittest.TestCase`
 def new_test(name, solver, import_dict, path):
     """logged_example_testcase with a NewDefaultSolver"""
     def test(self):
-        "Tests and logs stdout with a different default solver."
+        # pylint: disable=missing-docstring
+        # No docstring because it'd be uselessly the same for each example
```
Comment (whoburg) [inline gpkit/tests/helpers.py]: 👍

**Comment by whoburg** [comment]
review complete

File: gpkit/keydict.py (hunk 1 above)
Comment (bqpd) [inline gpkit/keydict.py]: internally only, should never be in a user-accessible dictionary.

File: gpkit/constraints/set.py (hunk 2 above)
Comment (bqpd) [inline gpkit/constraints/set.py]: this switch is actually a bugfix, in case k is a string...

File: gpkit/keydict.py (hunk 3 above)
Comment (bqpd) [inline gpkit/keydict.py]: Yeah. Right now it's really only for numpy arrays, but I figured I'd be a bit more general.

**Comment by bqpd** [comment]
Responded to review and added an inline comment.

File: gpkit/constraints/set.py (hunk 2 above)
Comment (whoburg) [inline gpkit/constraints/set.py]: got it

File: gpkit/keydict.py [hunk 5]
```diff
@@ -45,6 +45,7 @@ def update(self, *args, **kwargs):
         "Iterates through the dictionary created by args and kwargs"
         for k, v in dict(*args, **kwargs).items():
             if hasattr(v, "copy"):
+                # we don't want just a reference (for e.g. numpy arrays)
```
Comment (whoburg) [inline gpkit/keydict.py]: Was asking something a little more subtle here. If a user writes their own class (e.g. `pint.Quantity`, but possibly something else, what are the requirements for it to work as a value in a KeyDict?

 - Does it need to be immutable?
 - Does it need to supply a `copy()` method?

**Comment by whoburg** [comment]
I haven't thought hard about it yet, but on an initial read it's not obvious to me how the `"filter"` functionality works. I don't mean to hold up the merge but I think that needs to be documented (inline comment at the very least), since it seems it could potentially interact with user-defined constraint classes.

If merging this is super urgent we could make an issue out of that, but I imagine it just needs a few good inline comments to close that concern now and merge this.

File: gpkit/keydict.py (hunk 5 above)
Comment (bqpd) [inline gpkit/keydict.py]: Either one or the other!

**Comment by bqpd** [comment]
@whoburg, added inline comments to keydict and constraintset.subinplace, ready for review

File: gpkit/constraints/set.py [hunk 6]
```diff
@@ -174,10 +174,15 @@ def flat(self, constraintsets=True):
...
         for constraint in self:
+            # break down the subs dictionary to pass each subconstraint
+            # only the keys that are in it, skipping subconstraints
+            # which don't need substitutions at all
             if "filter" in subs:
+                # all keys are already varkeys
                 subs_ = {k: v for k, v in subs.items()
                          if dict.__contains__(constraint.varkeys, k)}
             else:
+                # turn keys into varkeys
```
Comment (whoburg) [inline gpkit/constraints/set.py]: this looks good, thanks (trying out the new github review system)

Comment (whoburg) [inline gpkit/constraints/set.py]: replying within this thread

**Comment by whoburg** [comment]
somewhat surprised that github instructed me to merge master into this branch. Did so though (via a seemingly new "merge master into this branch" button on the UI).

**Pull request merged**